liquidprompt

```diff
@@ -1133,6 +1134,31 @@ _lp_python_env_color() {
     lp_python_env_color="[${LP_COLOR_VIRTUALENV}${lp_python_env}${NO_COL}]"
 }
 
+_lp_ruby_env() {
+    (( LP_ENABLE_RUBY_VENV )) || return 2
+
+    local ret
+
+    if command -v rvm-prompt >/dev/null ; then
+        [ -z "${LP_RUBY_RVM_PROMPT_OPTIONS-}" ] && LP_RUBY_RVM_PROMPT_OPTIONS=(i v g s)
```

Comment:
This is a config option, so it should be set to a default after line 218.

Detecting an empty or not set array is tricky across all of the shell versions we support (see https://stackoverflow.com/a/61551944/8338372), see line 226 for an example of how to do it correctly. This is more complex, since if the array is not set, we want to give it a default value, not leave it as an empty array. I think this is what we should do:

On line 219, where setting the default:
```bash
    LP_RUBY_RVM_PROMPT_OPTIONS=( ${LP_RUBY_RVM_PROMPT_OPTIONS[@]+"${LP_RUBY_RVM_PROMPT_OPTIONS[@]}"} )
    [[ ${#LP_RUBY_RVM_PROMPT_OPTIONS[@]} == 0 ]] && LP_RUBY_RVM_PROMPT_OPTIONS=(i v g s)
```
And then on line 1144, where using it, we shouldn't need unset protection (but do need quotes):
```bash
        __lp_escape "$(rvm-prompt "${LP_RUBY_RVM_PROMPT_OPTIONS[@]}")"
```

Reply:
I’m really not a crack at shell expansion. But is there a difference between:

```shell
# Initializing the array with maybe external content, then testing against length of the array
 LP_RUBY_RVM_PROMPT_OPTIONS=( ${LP_RUBY_RVM_PROMPT_OPTIONS[@]+"${LP_RUBY_RVM_PROMPT_OPTIONS[@]}"} )
[[ ${#LP_RUBY_RVM_PROMPT_OPTIONS[@]} == 0 ]] && LP_RUBY_RVM_PROMPT_OPTIONS=(i v g s)
```
and

```shell
# Testing against the existence of the variable and setting it to the default value if not (as seen on line 301 for LP_COLORMAP)
[[ -z "${LP_RUBY_RVM_PROMPT_OPTIONS-}" ]] && LP_RUBY_RVM_PROMPT_OPTIONS=(i v g s)
```

The second seems more quick/readable than the first for me… and in the case of rvm prompt options we don’t want to merge anything, but just have default value or nothing. What do you think?

Reply:
Yes, there is a difference. The second will evaluate to true if the array has elements and the first element is the empty string. (The reason is that shell is dumb, and referencing an array var without the `[@]` syntax gets the first element of the array.) That is unlikely for this instance, but possible.

It is very likely on the color map instance though, and I'm surprised that no one has reported a bug there. I guess no one is customizing their color map; this is the second fatal bug I have seen on this feature. Good find, I'll be fixing that. You can open a Github issue if you want Github points :smile:

Reply:
I did some further digging, and it seems there is no way (supported by Bash 3.2+ and Zsh 5.0+) to get the number of elements in an array while being `set -u` safe. That is:
1. `${#var[@]}` does not work (errors when `var` is not set).
2. `${#var[@]-0}` does not work (syntax error).
3. `${var[@]+${#var[@]}}` does not work (returns nothing instead of `0` when the array is unset or empty).

Which leaves our only option to `var=( ${var[@]+"${var[@]}"} )` (setting the array to itself, or if it was unset, setting it to an empty array), because then we *know* it is set, and can then simply do a `[[ ${#var[@]} == 0 ]]` check to see if it is empty.

Reply:
Thank you for your explanations, and let’s do this.